Rewrite IItemHandler

loordgek opened the pull request:
this is my first PR :) 

this pr fixes https://github.com/MinecraftForge/MinecraftForge/issues/3344
by adding isStackValidForSlot and canExtractStackFromSlot.

and makes the IItehandler slotless(sort of) it can still be used slotted.
i dont see splitting it up into to 2 capabilitys as a option.

the method calcRedStoneFromInventory is there because  i had a problem w/ buildcraft gates and
storage drawers drawer w/ voiding upgrade installed telling the gate it was not full but it was voiding the rest, not useful for automation 

the Cabability for IItemHolder exits for use on ItemStacks

loordgek committed:
Rewrite IItemHandler

CLAassistant commented:
[![CLA assistant check](https://cla-assistant.io/pull/badge/signed)](https://cla-assistant.io/MinecraftForge/MinecraftForge?pullRequest=4630) <br/>All committers have signed the CLA.

loordgek commented:
the reason why i create this pr now and not wait for 1.13 is there are many things wrong w/ this pr 
so we can them fixed

williewillus commented:
general comments:
* If this is targeting 1.13, remove meta from all methods that have it, stack meta no longer exists.
* Can you briefly explain all the transaction stuff? Why is it needed and what does it make possible that isn't now?
* Can you explain (like you're teaching a new modder) the difference between IItemHandler and IItemHolder? Why are there two and which one should I use? (Is it worth the increased complexity in Forge to have two inventory interfaces)
* Do your additions properly work for vanilla IInventory and ISidedInventory?
* Can we have specified behaviour for what happens when concurrent modification occurs to the iterators?

liach commented on src/main/java/net/minecraftforge/items/IItemHandler.java:
```diff
@@ -19,73 +19,262 @@
 
 import javax.annotation.Nonnull;
 
 public interface IItemHandler
```
You should make this a list so that we can iterate over slots with a list iterator

loordgek replied:
does the IItemHandlerIterator work for you ?

liach commented on src/main/java/net/minecraftforge/items/IItemHandlerIterator.java:
```diff
@@ -0,0 +1,44 @@
+import javax.annotation.Nonnull;
+import java.util.Iterator;
+
+public interface IItemHandlerIterator extends Iterator<ItemStack>
```
A list iterator can serve this function, I believe.

loordgek replied:
i was planning to allow to skip Empty but i dint get that working 
like
default IItemHandlerIterator itemHandlerIterator(boolean skipEmpty)

and i dont want add/set/remove for that you can use https://github.com/loordgek/MinecraftForge/blob/4d2adefed21c2688a9635919610dbf9189eae410/src/main/java/net/minecraftforge/items/IItemHandler.java#L279

liach commented on src/main/java/net/minecraftforge/items/InsertTransaction.java:
```diff
@@ -0,0 +1,53 @@
+
+import javax.annotation.Nonnull;
+
+public class InsertTransaction
```
Final?

liach commented on src/main/java/net/minecraftforge/items/filter/ItemStackFilter.java:
```diff
@@ -0,0 +1,161 @@
+    private final ItemStack stack;
+    private final Item item;
+    private final Range<Integer> metadata;
+    private final NBTTagCompound nbtTag;
```
Should use a nbt predicate instead

liach commented on src/main/java/net/minecraftforge/items/filter/ItemStackFilter.java:
```diff
@@ -0,0 +1,161 @@
+    private final boolean matchStackSize;
+    private final boolean inverted;
+
+    public ItemStackFilter(ItemStack stack, Item item, Range<Integer> metadata, NBTTagCompound nbtTag,
```
Builder plz

liach replied:
Noop, if you have a builder leave this protected or private

liach commented on src/main/java/net/minecraftforge/items/filter/ItemStackFilter.java:
```diff
@@ -0,0 +1,161 @@
+        private ItemStack stack = ItemStack.EMPTY;
+        private Item item = null;
+        private Range<Integer> metadata = null;
+        private NBTTagCompound nbtTag = null;
```
Nullability annotations

liach commented:
Ah, my eyes are tired! Good effort. I need to look into it deeper.

loordgek commented:
@williewillus 

- gong to keep it in for now probably rename it damage 

- dont know really. need the hear from the enderio devs what they need probably a OptionalInt for the slot it extracted/inserted from

- the IItemHandler is used for interacting between mods, the IItemHolder is used for Container/Slot interacting, the is a Capability available for the IItemHolder for use on ItemStacks

- yes

- will work on that

loordgek committed:
register IItemHolder cap

make InsertTransaction final

add javadoc to IItemHandlerIterator

use Predicate for the NBTTagCompound in the ItemStackFilter

removed Storage forgot to delete it

liach commented on src/main/java/net/minecraftforge/items/CapabilityItemHandler.java:
```diff
@@ -42,41 +39,20 @@ public static void register()
-                }
-                return nbtTagList;
+                if (!(instance instanceof INBTSerializable))
+                    throw new RuntimeException();
```
an illegalargumentexception may be more clear

liach commented on src/main/java/net/minecraftforge/items/ItemHandlerHelper.java:
```diff
@@ -19,38 +19,30 @@
-        if (dest == null || stack.isEmpty())
-            return stack;
+    public final static IInventory emptyInventory = new InventoryBasic("[Null]", true, 0);
+    public final static EntityEquipmentSlot[] arrmorSlots = new EntityEquipmentSlot[]{
```
1. public static final modifier orders
2. typo arrmor, should be armor

liach commented on src/main/java/net/minecraftforge/items/ItemHandlerHelper.java:
```diff
@@ -197,36 +139,219 @@ public static void giveItemToPlayer(EntityPlayer player, @Nonnull ItemStack stac
+        return handler.getSlotLimit(slot);
+    }
+
+    public static void MultiExtract(IStackFilter filter, Range<Integer> slotRange, IExtractionManager manager, boolean simulate, IItemHandlerModifiable handler)
```
use lower camel case instead of upper camel case

liach commented on src/main/java/net/minecraftforge/items/ItemHandlerHelper.java:
```diff
@@ -197,36 +139,219 @@ public static void giveItemToPlayer(EntityPlayer player, @Nonnull ItemStack stac
+        if (amount == 0) return ItemStack.EMPTY;
+        if (isRangeSingleton(slotRange))
+            return extract(slotRange.lowerEndpoint(), filter, amount, simulate, handler);
         else
```
If you return, you don't need an else

liach commented on src/main/java/net/minecraftforge/items/IItemHandlerIterator.java:
```diff
@@ -0,0 +1,104 @@
+     * @return {@code true} if the list iterator has more elements when
+     *         traversing the list in the forward direction
+     */
+    boolean hasNext();
```
`@Override`

liach commented on src/main/java/net/minecraftforge/items/IItemHandlerIterator.java:
```diff
@@ -0,0 +1,104 @@
+     * @throws NoSuchElementException if the iteration has no next element
+     */
+    @Nonnull
+    ItemStack next();
```
`@Override`

liach commented on src/main/java/net/minecraftforge/items/IItemHandlerIterator.java:
```diff
@@ -0,0 +1,104 @@
+     *         subsequent call to {@code previous}, or -1 if the list
+     *         iterator is at the beginning of the list
+     */
+    int previousIndex();
```
More methods:
`int index();` (current index)
`TransactionResult set(ItemStack replacement)`

loordgek committed:
typos

more methods for IItemHandlerIterator

liach commented on src/main/java/net/minecraftforge/items/IItemHandlerIterator.java:
```diff
@@ -75,19 +76,52 @@
+     * @return the Remainder that was not inserted
+     */
+    @Nonnull
+    ItemStack set(@Nonnull ItemStack stack);
```
Can we have some sort of insertion result here?

loordgek replied:
like ??

liach commented on src/main/java/net/minecraftforge/items/IItemHandlerIterator.java:
```diff
@@ -75,19 +76,52 @@
+     * @return the ItemStack at the index it is currently at
+     */
+    @Nonnull
+    ItemStack stackAtIndex();
```
`current` is a better name

williewillus commented:
> gong to keep it in for now probably rename it damage

umm why? meta is completely gone, the concept of it no longer exists in 1.13. Tool damage etc. is moved into the NBT tag and there is already a flag to compare by NBT

> dont know really. need the hear from the enderio devs what they need probably a OptionalInt for the slot it extracted/inserted from

"I don't know" is not very convincing :/

> the IItemHandler is used for interacting between mods, the IItemHolder is used for Container/Slot interacting, the is a Capability available for the IItemHolder for use on ItemStacks

Why do they need to be separate? The current solutions (SlotItemHandler, validation at slot level for GUI, handler level for automation) seem to work just fine.

loordgek commented:
handy, shortcut. and when you compare NBT it just equals, i use range to compare so is the damage between 0 and 100

waiting for enderio devs

gigaherz and yueh liked the idea for separate interface but looking at it now we dont need it from what i can see

williewillus commented:
"handy shortcuts" is not a good reason to pollute the public interface with an idea that *no longer exists in 1.13*, I don't know how else I should state that.

yueh commented:
### Transactions

The `InsertTransaction` is probably really badly named and should something like `InsertResult`. The original issue here is probably that there is no way to tell, if an itemstack was destroyed or stored. It appears that it wants to solve it by making the leftover available. 

Nevertheless it does not really solve it. Something like an enderchest with distinct input and output variants will still break it an enforce that every monitored inventory needs to be rescanned once a single inventory changes. 

All it does is adding a convenient way to reassign the leftover amount instead of subtracting the used amount. But this comes at a high cost, nothing is enforced about their behaviour. Worst case is that someone passes the original itemstack as part of the itemstack, but modifies the size. Should someone still need the original one in an unmodified state, it will screw them up and force them to create a full copy before passing it along, as well as making an immediate copy for the return value in case it actually exposes an internal itemstack. This is actually an issue currently, as it enforces a very defensive approach to copy any itemstack returned by `ItemHandler`s, even if most mods actually return a safe object and thus the stack is copied twice. Which in high throughput situations is actually noticable.

### Using `Range`

It does not give any advantage, only disadvantages. It completely eliminates direct access and enforces everyone to iterate over their whole inventory until they find a matching slot. Same for `all()`, there is no way to tell it being all except some really broken hacks. Thus no way to handle it besides `iterate over the whole inventory`. Keeping the `int` and definining `-1` as wildcard is way better, even when still being bad.

It is the usual issue when using a compromise. The result is worse than sticking to either A or B.

### Handler/Holder

No idea about it. The holder is basically something for internal use. As long as forge does not provide a fully fledged GUI framework I do not see many reasons for it.

But I see the issues with `IItemHandler` lacking for GUI related tasks or some other specialised ones. E.g. `IItemHandlerModifiable`. `IIventory` is actually better for these tasks, but then also overkill as it has to be fully implemented should some other mod still favor it over itemhandlers.

### Multi extract.

Sounds nice as theory. But it will probably be either too limited for most use cases or way to complex.
Backtracking needs to be possible to extract multiple sets. Say someone wants to craft 12 fences and you want to extract 12 sets each containing 4 planks and 2 sticks. The inventory might have 48 planks, but only 4 sticks. Should the planks be extracted before the sticks, you still want to have a chance to backtrack and change it to only extract 8 planks. Sure you could jost loop 12 items. But that breaks immediately down once something requires a couple thousand stacks and each loop just extracts 1 cobble/plank.

Trying to solve every case will just make it way too complicated to implemented correctly. If not it will be very limited, but devs still have to potentially handle it, requiring additional effort for no benefit.

### `calcRedStoneFromInventory`

Really a bad idea. So many non inventories could make use of it for various reasons. So they either have to fake a `NullItemHandler` to abuse it. Or rely on vanilla methods. Which then requires patches for the comparator to support the itemhandler cap and the normal vanilla methods.

Extracting it to its own cap would still require patches, but any TE could make use of it without abusing some unrelated cap. Would make it way more flexible.


### Iterator

Also something potentially dangerous. Especially with all additional methods to traverse in both directions. Iterating over an inventory should be avoided as much as possible as it can be costly.

Outside of accessing a specific slot, no external user should have to scan over all slots until they find a matching one. The basic/generic approach should always be "give me X" and let the itemhandler figure out the least expensive way to do it. Slotted access is really something for UI or very special mod interactions, But these have by design more knowledge about that particular case and thus can decide to take a better path (that is the main reason for the handler/holder idea)

### No way to monitor inventories.

For us pretty much the most important feature. It would pretty much allow us to get rid of constant rescans of all inventories because a single one changes. 

Also it makes iterators mostly useless. Even simple pipes could just listen for changes and only wake up once the desired items enters the inventory. Instead of iterating over hundred slots each tick to check, if it is finally here. And it could be a simple as flipping a boolean once the item matches the configured filter. No need to keep a complete copy of the inventory as cache locally. (Should less memory usage be important).

An iterator would pretty much only be needed for an initial cache. But even that can be solved by defining that a listener will recieve all contents as delta after registering themselves to an observable inventory.

### Stuff I have forgotten. Or overlooked

Probably a bunch. Pretty large PR to miss something. And I do not remember every old idea I had.

raoulvdberge commented:
> No way to monitor inventories.

This is what we *really* need.

loordgek commented:
> the method calcRedStoneFromInventory is there because i had a problem w/ buildcraft gates and
storage drawers drawer w/ voiding upgrade installed telling the gate it was not full but it was voiding the rest, not useful for automation


> Using Range

i used int but ran into problem w/ the RangedWrapper, amadornes told me to use Range

> No way to monitor inventories.

what if somebody is not using IItemHandler(IC2)

raoulvdberge commented:
> what if somebody is not using IItemHandler(IC2)

Too bad.

yueh commented:
> the method calcRedStoneFromInventory is there because i had a problem w/ buildcraft gates and
storage drawers drawer w/ voiding upgrade installed telling the gate it was not full but it was voiding the rest, not useful for automation

That is something completely different. Either it is a mod issue. Then the itemhandler should not be broken to fix their stuff. Or it something way more advanced, which should not pollute the itemhandler and should be better in their own cap (like there is a need for more information based caps like `isBusy` or `isFull`. This might overlap with `ItemHandler`. But if something simply voids any overflow, so be it. In some cases it might be actually be a desired feature.

> i used int but ran into problem w/ the RangedWrapper, amadornes told me to use Range

Then the RangedWrapper should be fixed. Not choose the worst option for everything else just to avoid it.

> what if somebody is not using IItemHandler(IC2)

Ignore them. If a feature is a massive performance improvement and/or reduces liability (as in maintained code), it is easy to pay the cost of dropping support for a small amount of mods.

loordgek commented:
you can use Range.singleton for slotted handling

williewillus commented:
It would be unwise to just "ignore them" for inventories not using IItemHandler because there's an important mod that doesn't, vanilla.

loordgek commented:
vanilla gets redirected to the itemHandler so that is not the problem

yueh commented:
> you can use Range.singleton for slotted handling

No you can't. Because it only allows to iterate over the whole inventory and match the current slot against the range until it matches. Sure you can argue that a vanilla furnace has only 3 slots and that wouldn't be soo bad. But you can't design around a vanilla furnace. You have to take into account that someone might add an inventory with 1 million slots. Sure it's a pretty bad idea. But it is only because the external user should try to find the slot to insert the stack instead of letting the inventory deal with it. Which might be optimised to use a hashed key using the item id.

Sure you could just ignore the range internally and do your own thing. Until some other mod fails because they require it to be used and starts to dupe stuff.

Actually it does not make any use of a range at all. E.g. by making different ranges of slots available by their use. Say `[0,3)` for inputs, `[3,4)` output and `[4,10]` upgrades/optional stuff. There is simply no way to query for their type.
Essentially the range just tries to imitate an `Optional<Integer>`, with `absent` being abused as wildcard. Or even more precise a `Predicate<Integer>`. Just add all downsides of passing a predicate around, while immediately preventing any benefit from being usable. (Even if these are very special cases and usually limited to mod internal use or very tightly coupled mods)

loordgek commented:
> Because it only allows to iterate over the whole inventory and match the current slot against the range until it matches

no i use this https://github.com/loordgek/MinecraftForge/blob/9c04dc7b0d1adbc29506ab9b498d34397acbc87e/src/main/java/net/minecraftforge/items/ItemHandlerHelper.java#L241-L243

yueh commented:
Yes, I know you can deliberately choose to break basic OOP principles. But that doesn't make it a valid point. If you need different code paths depending on the type of object, there is something wrong.

That function is used pretty much everywhere, which is like adding "OK, we know the design is bad and in 99.9% you'll want to use a special case to avoid performance issues. But we need it for one rare bug. Read more here." to the docs. If that is a not completely unwarranted, it is a pretty big hint that there is something wrong with the design.

It is essentially offloading a single issue with `RangedWrapper` onto every other dev, force them to be aware of performance issues as well as explicitly handle it to avoid these.
But that is simply not feasible. Not every dev will be fully aware of the issues it could cause or depending on their implementation even be able to avoid it. The basic `IItemHandler` should be as simple as possible, without giving anyone a chance to mess something up. Either the implementation or the external user by doing something probelematic like unecessary iterations over the full inventory.

Would the guava `Range` be a bit more like Python and work as generator, it would be easy to just iterate over the range and each loop would work with a concrete integer index. Not perfect, but at least the code is identical for an empty range, singleton or any other variant.

loordgek commented:
then for the rangedWrapper i need iterate over the slots for slotless mode.
do you think that is a problem

yueh commented:
Hard to say without any statistics about it even being used.

But the real issue is still that trying to shove slotless and slotted interaction into the same cap will not work. They are simply too disjunct for it and any compromise will have some downsides. Certainly splitting it into two separate ones won't be ideal, too. But providing default wrappers shouldn't be that hard. Wrapping a slotted inventory into a slotless one is essentially just iterating over all slots. The other way is more or less wrapping a slotless into a slotted with 1 slot or as many as different stacks are stored. This will probably cost a bit more effort to handle correctly for every case.

Should it be impossible to introduce multiple variants depending on the use case, than a compromise should enforce as few limitations as possible.
E.g. while a `RangedWrapper` would need to iterate over them for a slotless access, in most cases it is probably using only a couple of slots. Which would not be that bad. Still it would not prevent anyone from using a specialised implementation without having to cover some quirky workarounds due to using `Range`.

Also the `IItemHolder` pretty much cover the slotted approach. So it basically adds two slightly different slotted caps on top of any `IInventory` variant. Thus a couple of redundant caps covering the same use cases, but still no real slotless one.

Ultimately the main performance issue is still being unable to monitor an inventory for changes in an efficient way and a single change stack essentially invalidates every other existing inventory. Which leads to a single write operation easily leading to hundreds of reads (or even more). Which are useless in 99.9% of all cases, but that one inventory has to break it.


What I did miss is the `IStackFilter`. Especially in combinations with the ignore NBT (or obsolete ignore meta). These things are very mod specific and then do not even cover all use cases. Potentially making them even impossible to achieve. For example a perfectly valid use case is to a any damage value (the real deal. not metadata), but match a specific NBT tag. Which would be impossible to achieve as damage will be an arbitrary nbt tag and with no partial matching it will never match any damage (or at least not without testing every existing itemstack against the filter).
Similar for the other way. Specific damage (or range) is not possible without also enforcing a full NBT compare, which is way more costly than a simple integer compare.

IMO the whole PR currently tries to satisfy too many mod specific approaches, which ultimately just adds potential pitfalls all over the place.

loordgek committed:
rewrite range to OptionalInt

removed a bunch on stuff

loordgek commented:
just did apply your suggestions how does it look now

williewillus commented on src/main/java/net/minecraftforge/items/IItemHandler.java:
```diff
@@ -19,73 +19,154 @@
+     * @param slot  to check
+     * @return true is the stack can be inserted into the slot
+     */
+    default boolean isStackValidForSlot(@Nonnull ItemStack stack, int slot)
```
reverse the arguments; convention has always been slot args first when they are present

williewillus commented on src/main/java/net/minecraftforge/items/IItemHandler.java:
```diff
@@ -19,73 +19,154 @@
-     *         May be the same as the input ItemStack if unchanged, otherwise a new ItemStack.
-     *         The returned ItemStack can be safely modified after.
-     **/
+     * @param slotRange use {@link Range#all()} for slotLess extraction, use {@link Range#singleton(Comparable)} for slotted extraction
```
update your javadocs

williewillus commented on src/main/java/net/minecraftforge/items/IItemHandlerObserver.java:
```diff
@@ -0,0 +1,12 @@
+
+public interface IItemHandlerObserver
+{
+    void onStackInserted(IItemHandler handler, @Nonnull ItemStack oldStack, @Nonnull ItemStack newStack, int slot);
```
how does this handle a case where only the stack count changed?
also move slot argument to be the second one

loordgek replied:
yes

williewillus replied:
> how

what would old and newstack be if I insert 1 stone into an inventory and it stacks with an already existing stack of 1 stone? Wouldn't this force unnecessary copies if only the stack size changed? Because you could've just called `grow(1)` on the one in the inventory.

loordgek replied:
yes, is the oldstack usefull to have?

williewillus commented on src/main/java/net/minecraftforge/items/ItemHandlerHelper.java:
```diff
@@ -197,36 +138,126 @@ public static void giveItemToPlayer(EntityPlayer player, @Nonnull ItemStack stac
-     * @return A redstone value in the range [0,15] representing how "full" this inventory is.
-     */
-    public static int calcRedstoneFromInventory(@Nullable IItemHandler inv)
+    public static int getFreeSpaceForSlot(IItemHandler handler, int slot)
```
why is this not just a method on IItemHandler

(or any of these, for that matter)

loordgek replied:
had it in there, will move it back

williewillus commented on src/main/java/net/minecraftforge/items/ItemHandler.java:
```diff
@@ -0,0 +1,118 @@
+import java.util.List;
+import java.util.OptionalInt;
+
+public class ItemHandler implements IItemHandlerModifiable, INBTSerializable<NBTTagCompound>
```
no need to rename from ItemStackHandler to this

williewillus commented on src/main/java/net/minecraftforge/items/filter/IStackFilter.java:
```diff
@@ -0,0 +1,39 @@
+    @Override
+    boolean test(ItemStack stack);
+
+    default NonNullList<ItemStack> getExamples()
```
what use is this?

loordgek replied:
can be used in a indexed inventory https://github.com/CyclopsMC/CyclopsCore/blob/master-1.12/src/main/java/org/cyclops/cyclopscore/inventory/IndexedInventory.java

williewillus replied:
then it needs a better name and full javadoc stating what it's supposed to be used for, its guarantees about the returned list and stacks (mutable? immutable?), etc.

williewillus commented on src/main/java/net/minecraftforge/items/wrapper/RangedWrapper.java:
```diff
@@ -19,95 +19,150 @@
+    private final IItemHandler compose;
+    private final int min;
+    private final int maxSlotExclusive;
+    protected final Range<Integer> scanrange;
```
`scanRange`

yueh commented:
Just a short notice after a first glance.

The observer part probably needs some thoughts. Not this particular approach, but in general.
E.g. Caps should not be cached, which makes listeners a bit problematic in various aspects. For example there might be no way to link a observed handler to its TE. This might make handling chunk unload a bit difficult as an `IItemHandlerObserver` cannot be notified once it is invalidated.

Further it needs some contracts about how registering an `IItemHandlerObserver` is handled. Are they notified about the inital inventory? Or does it have to be fetched via the normal inventory? Either case needs to be explicit, otherwise it will run into invalid caches. Similar for unloading an inventory. Does it unroll everything by sending an update for every slot being set to empty. Or just notify it about being invalidated? The second approach will pretty much make it necessary to keep a local cache for each observer should it require to forward changes. The first one might allow it to be avoided, but the question is how many mods will implement it incorrectly.

Also the other way with an observer becoming invalid, if the forget to perform an explicit unregister, it could easily lead to leaking old TEs.

loordgek commented:
https://gist.github.com/loordgek/72fe31857f620b249d8aac87875bf954
if we are going w/ this where to put WeakReferenceList ?

raoulvdberge commented:
Can't we just have an event instead of a error prone complicated observer system?

williewillus commented:
that doesn't solve most of the problems @yueh brings up though, specifically regards to:

* If the handler is in the event, is it still valid? (its containing TE/Entity's chunk could have unloaded so further changes aren't saved)
* How batch ops are handled (see above)

yueh commented:
An event would also be very awkward. The normal forge event bus is a global system, so each change would bubble up, each mod has their own listener to bubble it back down to whatever actually monitors the inventory just to bubble the change back up to any parent monitor.

Which in no way deals by default with cases like the monitor being invalidated. Efficient delegation is probably not something forge can provide by default and is up to the actual mod. Further it does not solve the common issues with observers and over/undercommunicating. These still exist with an eventbus. Say "But I was under the impression that you should send a simulate around so anyone could cancel it" introduces their own sorts of issues in any case. Nothing which could be even dealt outside mentioning it in the docs to never do it.

Or even end up with hundreds of event listeners registered to a single event bus and 99.5% just discard the event immediately due to being unrelated. This might work for a few hoppers, but I use testsetups, which can easily push 100k items/tick around. Which usually has to be at least split based on the max stacksize. So still a few thousand events/tick. They are even somewhat redundant. Say using storage drawers as source and target. In this case the source drawer and source controller has to send their own events as they cannot assume someone can detect that each belong to the same group. Or simply the drawer and controller are monitored separately. Same for the target, which just multiplied the amount of events by 4 and there is no real limit to it. Without an actual chance to suppress the event, should there be no registered listener. As the majority of inventories are probably unmonitored, it can easily oversaturate the bus with useless events.

I would even consider splitting it into its own cap. If at least the larger inventory mods support it explicitly, it can reduce the load quite a bit and change it from "Have to revalidate the whole world" to "Have to revalidate a few unknown inventories"
This also makes it easier to break the contract of being uncachable and just introduce an explicit `Listener#onInvalidate()` or similar.
Further reduce the interfaces to a bare minimum, probably without any default implementation as a concrete implementation will certainly differ. Just `Observer#add/removeListener()` and a generic 'onStackChanged`. When dealing with deltas, a separate method for extract/insert is probably not necessary. Futher a `Listener#isValid()` might be useful to remove an old listener. Probably safer than relying on the GC to remove it implicitly, as there are no guarantees when it happens. Of course also a `Listener#onObserverableInvalidated()` or similar.

loordgek commented:
yueh i like the cap, idea for sided inves expose the observer on the same side ?
but i am keeping extract/insert handler slot stack for use that a Observer can see the slot and if they need that stack handler#extract(bladiebladiebla)

loordgek commented:
i have to provide a default implementation

loordgek committed:
added IItemHandlerObservable capability

loordgek commented:
added IItemHandlerObservable

yueh commented on src/main/java/net/minecraftforge/items/itemhandlerobserver/IItemHandlerObservable.java:
```diff
@@ -0,0 +1,19 @@
+
+    void removeObserver(IItemHandlerObserver observer);
+
+    boolean containsObserver(IItemHandlerObserver observer);
```
This might be a bit superfluous. An external user should not have to deal with registering duplicated listeners or removing non nonexistent ones. Just have add/remove ignore these cases. And document that duplicate observer instances are not allowed.

yueh commented on src/main/java/net/minecraftforge/items/itemhandlerobserver/IItemHandlerObserver.java:
```diff
@@ -0,0 +1,17 @@
+
+    void onObserverableInvalidated();
+
+    boolean isNotValid();
```
`isNot` usually sounds somewhat strange, especially when having to invert it and deal with double negatives. `isValid` should be enough.

Also choosing another synonyms might be an idea to avoid confusion about it. As this is about the observer still being valid, while the `onObserverableInvalidated` is about the observed inventory.

yueh commented on src/main/java/net/minecraftforge/items/itemhandlerobserver/IItemHandlerObserver.java:
```diff
@@ -0,0 +1,17 @@
+{
+    void onStackInserted(IItemHandler handler, int slot, @Nonnull ItemStack newStack);
+
+    void onStackExtracted(IItemHandler handler, int slot, @Nonnull ItemStack newStack);
```
Adding the oldstack or new/old amount might still have it's benefits. It avoids the need of using another local cache, should an observer need to forward the changed amounts as delta.

loordgek committed:
cleanup

liach commented on src/main/java/net/minecraftforge/items/itemhandlerobserver/IItemHandlerObserver.java:
```diff
@@ -7,11 +7,11 @@
+    void onStackExtracted(IItemHandler handler, int slot, @Nonnull ItemStack oldStack, @Nonnull ItemStack newStack);
 
-    void onObserverableInvalidated();
+    void onObservableeInvalidated();
```
`onInvalidation`

loordgek committed:
onObservableeInvalidated > onInvalidated

loordgek commented:
are we good now ??

loordgek commented:
https://gist.github.com/loordgek/ea47633b2c9fe2e5a3ffd2e8969aae63
can be handy for anti steal mods
you can also set it that a inv can only be extracted with a hopper

loordgek committed:
add ItemTransferEvent

and javadoc

loordgek committed:
add ItemTransferEvent

and javadoc

loordgek committed:
typos

loordgek committed:
typos

yueh commented:
With 1.13 not having an imminent release, there is still time left to fetch further feedback. I personally cannot imagine every possible use case. Worst case might be that it gets merged for 1.13, some devs do not release it and only request changes after the first recommended build once they start porting. Which might make API changes very unlikely and screw them until 1.14 or later.


A (cancelable?) event is probably a really bad idea. It will certainly be the most common event, easily a few orders of magnitude larger. Take a `ItemCraftedEvent` for example. Which can easily use 10 items (9 inputs, 1 output), which all need to be extracted and then put into the machine or final storage. Which can easily trigger 20 events. 

Should they be cancelable, you have to deal with some very strange situations, E.g. not being able to insert it into the machine, but also not longer be able to resolve the issue by putting the items back into their storage. You might have to drop them onto the ground or void them as last restort. Both are very bad ideas. The first for performance reason, having thousands of items floating around is just bad. The other is really bad for a player once they see their items got voided.

Also the `ItemCraftedEvent` can already be a massive performane issue due to some suboptimal decisions. E.g. someone has the idea to query every event against a database. Even if it just takes a few μs per transfer, it can easily exceed the available ticktime. It is simply not like most other events either limited by the player not being able to click 100.000 times a second or they just happen once in the lifetime of a longer living entity/object.

yueh commented on src/main/java/net/minecraftforge/items/wrapper/EntityEquipmentInvWrapper.java:
```diff
@@ -104,12 +104,9 @@ public void clearInv()
     @Override
-    @Nonnull
-    public ItemStack setStackInSlot(int slot, @Nonnull ItemStack stack)
+    public void setStackInSlot(int slot, @Nonnull ItemStack stack)
```
Returning the previous stack is actually pretty useful.

One of the main issues with using `IItemHandler` for GUI interactions, is that there is no atomic way to swap a stack. Only use extract and insert and hope it does not get messed up somehow. Or us the `IItemHandlerModifiable`, which also has it's design issues. Like some helpers doing unchecked casts for it, Causing crashes or at least no functionality when used like documented.

Yes, you can combine `getStackInSlot` or similar, but this is stil not automatic. There are no guarantees that the handler does not modify the stack after using `setStackInSlot`, so it requires and explicit copy each time. Just returning the previous stack (with out any reference to some internal stack), would make swapping a slot very trivial. Pretty much no needs for safeguards. And it can be even used for fakeslots by simply returning `ItemStack.EMPTY` without any additional safeguards.

loordgek replied:
setStackInSlot is in IItemHandlerModifiable

yueh commented on src/main/java/net/minecraftforge/items/wrapper/IInvWrapperBase.java:
```diff
@@ -56,13 +56,10 @@ public boolean isStackValidForSlot(int slot, @Nonnull ItemStack stack)
     @Override
-    @Nonnull
-    public ItemStack setStackInSlot(int slot, @Nonnull ItemStack stack)
+    public void setStackInSlot(int slot, @Nonnull ItemStack stack)
```
Same here.

yueh commented on src/main/java/net/minecraftforge/items/ItemHandler.java:
```diff
@@ -38,13 +38,11 @@ public ItemHandler(int size)
+
     @Override
-    public ItemStack setStackInSlot(int slot, @Nonnull ItemStack stack)
+    public void setStackInSlot(int slot, @Nonnull ItemStack stack)
```
And here.

yueh commented on src/main/java/net/minecraftforge/items/itemhandlerobserver/IItemHandlerObserver.java:
```diff
@@ -7,11 +26,29 @@
+     * @param oldStackSize the old stackSize, 0 if it was {@link ItemStack#EMPTY}
+     * @param newStack the stack that is inserted into the slot
+     */
+    void onStackInserted(IItemHandler handler, int slot, int oldStackSize, @Nonnull ItemStack newStack);
```
How does this deal with the itemstack itself being changed? There is no way of telling, if the item itself changed. At least not without keeping a secondary cache around just to compare against.

yueh commented on src/main/java/net/minecraftforge/items/itemhandlerobserver/IItemHandlerObserver.java:
```diff
@@ -7,11 +26,29 @@
+     * @param oldStackSize the amount from the old stack
+     * @param newStack the new stack in the slot
+     */
+    void onStackExtracted(IItemHandler handler, int slot, int oldStackSize, @Nonnull ItemStack newStack);
```
Same as `onStackInserted`

loordgek commented:
> It will certainly be the most common event, easily a few orders of magnitude

that will still be entity tick events 

> Should they be cancelable

yes
the event is intended for anti grief/protection mods
we may need a multi insert/extract for IItemHandler and events for it

and BC handles this already https://github.com/BuildCraft/BuildCraftAPI/blob/8.0.x/api/buildcraft/api/transport/pipe/PipeEventItem.java#L40-L42

for the AE2 storage bus it should not a problem because i can reject the stack you are trying to put in the handler
